File: src/main/java/sonia/scm/statistic/Statistics.java
/*
 * Copyright (c) 2020 - present Cloudogu GmbH
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

package sonia.scm.statistic;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import sonia.scm.repository.Changeset;
import sonia.scm.repository.Modifications;
import sonia.scm.repository.Repository;
import sonia.scm.repository.api.RepositoryService;

import java.io.IOException;

public class Statistics implements AutoCloseable {

  private static final Logger LOG = LoggerFactory.getLogger(StatisticData.class);

  private StatisticManager statisticManager;
  private RepositoryService repositoryService;
  private StatisticData statisticData;

  Statistics(StatisticManager statisticManager, RepositoryService repositoryService, StatisticData statisticData) {
    this.statisticManager = statisticManager;
    this.repositoryService = repositoryService;
    this.statisticData = statisticData;
  }

  public RepositoryService getRepositoryService() {
    return repositoryService;
  }

  public Repository getRepository() {
    return repositoryService.getRepository();
  }

  public StatisticData getStatisticData() {
    return statisticData;
  }

  public void add(Changeset c) throws IOException {
    Modifications modifications = repositoryService.getModificationsCommand().revision(c.getId()).getModifications();
    statisticData = statisticData.add(c, modifications);
  }

  public void commit() throws IOException {
    statisticManager.store(this);
  }

  @Override
  public void close() {
    repositoryService.close();
  }
}

File: src/main/js/hook/statistics.ts
/*
 * Copyright (c) 2020 - present Cloudogu GmbH
 *
 * This program is free software: you can redistribute it and/or modify it under
 * the terms of the GNU Affero General Public License as published by the Free
 * Software Foundation, version 3.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see https://www.gnu.org/licenses/.
 */

import { apiClient } from "@scm-manager/ui-api";

export function getLinksForStatistics(url: string) {
  return apiClient
    .get(url)
    .then(response => response.json())
    .then(collection => {
      return collection._links;
    })
    .catch(err => {
      return {
        error: err
      };
    });
}

export function getCommitsPerAuthor(url: string) {
  return apiClient
    .get(url)
    .then(response => response.json())
    .then(collection => {
      return collection.author;
    })
    .catch(err => {
      return {
        error: err
      };
    });
}

export function getCommitsPerHour(url: string) {
  return apiClient
    .get(url)
    .then(response => response.json())
    .then(collection => {
      return collection.hour;
    })
    .catch(err => {
      return {
        error: err
      };
    });
}

export function getCommitsPerMonth(url: string) {
  return apiClient
    .get(url)
    .then(response => response.json())
    .then(collection => {
      return collection.month;
    })
    .catch(err => {
      return {
        error: err
      };
    });
}

export function getCommitsPerYear(url: string) {
  return apiClient
    .get(url)
    .then(response => response.json())
    .then(collection => {
      return collection.year;
    })
    .catch(err => {
      return {
        error: err
      };
    });
}

export function getCommitsPerWeekday(url: string) {
  return apiClient
    .get(url)
    .then(response => response.json())
    .then(collection => {
      return collection.weekday;
    })
    .catch(err => {
      return {
        error: err
      };
    });
}

export function getFileModificationCount(url: string) {
  return apiClient
    .get(url)
    .then(response => response.json())
    .then(collection => {
      return collection.modification;
    })
    .catch(err => {
      return {
        error: err
      };
    });
}

export function getTopModifiedFiles(url: string) {
  return apiClient
    .get(url)
    .then(response => response.json())
    .then(collection => {
      return collection.file;
    })
    .catch(err => {
      return {
        error: err
      };
    });
}

export function getTopWords(url: string) {
  return apiClient
    .get(url)
    .then(response => response.json())
    .then(collection => {
      return collection.word;
    })
    .catch(err => {
      return {
        error: err
      };
    });
}

export function rebuildStatistics(url: string) {
  return apiClient
    .post(url)
    .then(() => {
      return {
        success: true
      };
    })
    .catch(err => {
      return {
        error: err
      };
    });
}
